### ISSUE TITLE
🐞 Window doesnt move like the preview suggests

### ISSUE THREAD

**Issue by T1mVo** [MrKai77/Loop#462]
### Bug Description

When moving a window via the keyboard, it does not move as the preview suggests.

### Steps To Reproduce

1. Bind the move actions to some keys
2. Move a window

### Expected Behavior

The window should move into the exact same position as the preview.

### Actual Behavior

The window only roughly moves to the position shown by the preview.

### Screenshots

https://github.com/MrKai77/Loop/assets/94936637/6766dd2a-b991-40a7-be14-44226d588db1

### MacOS Version

Sonoma 14.2

### Loop Version

Version 1.0.0 (1030)

### Final Checks

- [X] My issue is written in English
- [X] My issue title is descriptive
- [X] This is a single bug (multiple bugs should be reported individually)

**Comment by MrKai77** [comment]
That is very interesting behavior. Do you have any external screens connected?

**Comment by T1mVo** [comment]
Yes, the recording was made on a 32:9 monitor. Though the problem also exists on the built-in screen.

**Comment by MrKai77** [comment]
Can you reproduce this bug when the external monitor is not plugged in?

**Comment by T1mVo** [comment]
> Can you reproduce this bug when the external monitor is not plugged in?

Yes.

---

Some things i observed:

- The window goes too far into direction it got moved, e. g. If I move the window to the left, it overshoots the preview to the left.
- Only the last changed direction affects the prior issue, e. g. If I move the window to the top and then to the right, it only overshoots to the right, not to the top.

**Comment by MrKai77** [comment]
Alright. Can you send me the file located at: `~/Library/Preferences/com.MrKai77.Loop.plist`?

**Comment by T1mVo** [comment]
Sure, sent via email.

**Comment by MrKai77** [comment]
Awesome, I will take a look! In the meanwhile, can you enable window animations in the Advanced tab, and see if Loop will behave differently compared to if it were off?

**Comment by T1mVo** [comment]
Its actually on already. First thing I tried was to turn it off as it is still under development. Issue still exists though.

**Comment by MrKai77** [comment]
I also just loaded your prefs into my Loop, and it seems to be working fine. Can you reproduce this with non-terminal-related apps as well?

**Comment by T1mVo** [comment]
Yes, happens with other applications as well.

**Comment by MrKai77** [comment]
Can you try running the latest [prerelease](https://github.com/MrKai77/Loop/releases/tag/prerelease)?

**Comment by T1mVo** [comment]
Tried it, no change

**Comment by stonevil** [comment]
Have the same issues with moving window. Issue become even more serious with external monitor. Loop show preview border window on "Next Monitor" but window itself moves only to the edge (right in this case) of the current screen. This happens only after 3-4 successful movement of window between two display.

MacBook Pro 16" 2019
External display: Samsung 28" Odyssey G70A
macOS Sonoma 14.5 (23F79)

**Comment by T1mVo** [comment]
In 1.1.0 and 1.2.0 this is now completely broken. The windows just gets moved completely to the left so that only a small portion is left on the screen.

**Comment by MrKai77** [comment]
Try this build: [Loop.zip](https://github.com/user-attachments/files/17081317/Loop.zip)
It should not fix anything, but I have added additional logging functionality to the app.

Once you run this app, open a Terminal, and run this command: 
```sh
log stream --predicate 'process == "Loop"' --style compact
```
This command will start logging everything that is happening to Loop while you use it.
Once this command is running, try looping a few times, then send me the logs. Hopefully this will help me pinpoint the problem!

**Comment by T1mVo** [comment]
Sure, here is the output. I moved the window to the right a couple of times and then released the modifiers.

```
log stream --predicate 'process == "Loop"' --style compact
Filtering the log data using "process == "Loop""
Timestamp               Ty Process[PID:TID]
2024-09-21 00:52:50.856 A  Loop[67216:268d39] (TCC) TCCAccessRequest() IPC
2024-09-21 00:52:50.857 Df Loop[67216:268d39] [com.apple.xpc:connection] [0x6000030f81e0] activating connection: mach=true listener=false peer=false name=com.apple.tccd.system
2024-09-21 00:52:50.874 Df Loop[67216:268d39] [com.apple.xpc:connection] [0x6000030f81e0] invalidated after the last release of the connection object
2024-09-21 00:52:50.878 A  Loop[67216:268d39] (TCC) TCCAccessRequest() IPC
2024-09-21 00:52:50.878 Df Loop[67216:268d39] [com.apple.xpc:connection] [0x6000030ee580] activating connection: mach=true listener=false peer=false name=com.apple.tccd.system
2024-09-21 00:52:50.886 Df Loop[67216:268d39] [com.apple.xpc:connection] [0x6000030ee580] invalidated after the last release of the connection object
2024-09-21 00:52:50.952 Df Loop[67216:268d39] [com.apple.AppKit:Window] order window: 448c op: 1 relative: 0 related: 0
2024-09-21 00:52:51.598 Df Loop[67216:268d39] (Foundation) Window action changed: moveRight
2024-09-21 00:52:51.599 Df Loop[67216:268d39] (Foundation) performKeybind: returning true due to valid event: moveRight
2024-09-21 00:52:51.604 Df Loop[67216:268d39] (Foundation) New preview window action received: moveRight
2024-09-21 00:52:51.766 Df Loop[67216:268d39] (Foundation) Window action changed: moveRight
2024-09-21 00:52:51.766 Df Loop[67216:268d39] (Foundation) performKeybind: returning true due to valid event: moveRight
2024-09-21 00:52:51.768 Df Loop[67216:268d39] (Foundation) New preview window action received: moveRight
2024-09-21 00:52:51.913 Df Loop[67216:268d39] (Foundation) Window action changed: moveRight
2024-09-21 00:52:51.913 Df Loop[67216:268d39] (Foundation) performKeybind: returning true due to valid event: moveRight
2024-09-21 00:52:51.914 Df Loop[67216:268d39] (Foundation) New preview window action received: moveRight
2024-09-21 00:52:52.081 Df Loop[67216:268d39] (Foundation) Window action changed: moveRight
2024-09-21 00:52:52.082 Df Loop[67216:268d39] (Foundation) performKeybind: returning true due to valid event: moveRight
2024-09-21 00:52:52.084 Df Loop[67216:268d39] (Foundation) New preview window action received: moveRight
2024-09-21 00:52:52.226 Df Loop[67216:268d39] (Foundation) Window action changed: moveRight
2024-09-21 00:52:52.226 Df Loop[67216:268d39] (Foundation) performKeybind: returning true due to valid event: moveRight
2024-09-21 00:52:52.231 Df Loop[67216:268d39] (Foundation) New preview window action received: moveRight
2024-09-21 00:52:52.365 Df Loop[67216:268d39] (Foundation) Window action changed: moveRight
2024-09-21 00:52:52.365 Df Loop[67216:268d39] (Foundation) performKeybind: returning true due to valid event: moveRight
2024-09-21 00:52:52.367 Df Loop[67216:268d39] (Foundation) New preview window action received: moveRight
2024-09-21 00:52:53.331 Df Loop[67216:268d39] (Foundation) Resizing Arc to moveRight on LS49A950U
2024-09-21 00:52:53.337 Df Loop[67216:268d39] (Foundation) Target window frame: (288.24444580078125, 35.0, 1695.0, 1395.0), Screen Frame: (0.0, 25.0, 5120.0, 1415.0)
2024-09-21 00:52:53.576 Df Loop[67216:268d39] [com.apple.AppKit:Window] window NSPanel 448c finishing close
2024-09-21 00:52:53.581 Df Loop[67216:268d39] [com.apple.AppKit:Window] order window: 448c op: 0 relative: 0 related: 0
2024-09-21 00:52:53.614 Df Loop[67216:26925e] (Foundation) Window resized to (-1685.0, 25.0, 1695.0, 1395.0)
```

**Comment by MrKai77** [comment]
@T1mVo Thanks! There most definitely seems like something is wrong in the window resize step.. and possibly the size calculation too 🤔

Just making sure, your settings are the exact same from when you sent me the `.plist` file above right? Meaning that there is 10px of padding on each side of the window?

**Comment by T1mVo** [comment]
i have tried it again. This time with and without padding. While deactivating the padding fixes the window moving all the way to the left, it still has the issues I have described in the beginning of this issue.

**With padding set to 10px:**

```
Filtering the log data using "process == "Loop""
Timestamp               Ty Process[PID:TID]
2024-09-21 12:16:27.091 Df Loop[67216:268d39] [com.apple.AppKit:Window] order window: 466b op: 1 relative: 0 related: 0
2024-09-21 12:16:29.307 Df Loop[67216:268d39] (Foundation) Window action changed: moveRight
2024-09-21 12:16:29.308 Df Loop[67216:268d39] (Foundation) performKeybind: returning true due to valid event: moveRight
2024-09-21 12:16:29.310 Df Loop[67216:268d39] (Foundation) New preview window action received: moveRight
2024-09-21 12:16:33.194 Df Loop[67216:268d39] (Foundation) Resizing Mail to moveRight on Built-in Retina Display
2024-09-21 12:16:33.201 Df Loop[67216:268d39] (Foundation) Target window frame: (540.4984130859375, 266.0, 878.0, 681.0), Screen Frame: (0.0, 44.0, 1800.0, 1125.0)
2024-09-21 12:16:33.443 Df Loop[67216:268d39] [com.apple.AppKit:Window] window NSPanel 466b finishing close
2024-09-21 12:16:33.445 Df Loop[67216:268d39] [com.apple.AppKit:Window] order window: 466b op: 0 relative: 0 related: 0
2024-09-21 12:16:33.495 Df Loop[67216:286df6] (Foundation) Window resized to (-868.0, 44.0, 878.0, 681.0)
```

**Without padding:**

```
Filtering the log data using "process == "Loop""
Timestamp               Ty Process[PID:TID]
2024-09-21 12:15:21.011 Df Loop[67216:268d39] [com.apple.AppKit:Window] order window: 4664 op: 1 relative: 0 related: 0
2024-09-21 12:15:22.327 Df Loop[67216:268d39] (Foundation) Window action changed: moveRight
2024-09-21 12:15:22.328 Df Loop[67216:268d39] (Foundation) performKeybind: returning true due to valid event: moveRight
2024-09-21 12:15:22.330 Df Loop[67216:268d39] (Foundation) New preview window action received: moveRight
2024-09-21 12:15:23.286 Df Loop[67216:268d39] (Foundation) Resizing Mail to moveRight on Built-in Retina Display
2024-09-21 12:15:23.293 Df Loop[67216:268d39] (Foundation) Target window frame: (540.4984130859375, 266.0, 878.0, 681.0), Screen Frame: (0.0, 44.0, 1800.0, 1125.0)
2024-09-21 12:15:23.543 Df Loop[67216:268d39] [com.apple.AppKit:Window] window NSPanel 4664 finishing close
2024-09-21 12:15:23.545 Df Loop[67216:268d39] [com.apple.AppKit:Window] order window: 4664 op: 0 relative: 0 related: 0
2024-09-21 12:15:23.586 Df Loop[67216:286540] (Foundation) Window resized to (540.0, 266.0, 878.0, 681.0)
```

**Comment by T1mVo** [comment]
I think Ive got the problem. The `bounds.size`'s in `WindowEngine.swift:101` are negative. In case of a padding of 10px the sizes are both -20, twice the padding.

Later in the animation the bounds check will produce the large negative value we are seeing in my last log.

```swift
let xDiff = lastWindowFrame.width - newFrame.width
if newFrame.maxX + xDiff > lastWindowFrame.maxX || currentValue >= 0.5,
   newFrame.maxX + xDiff > bounds.maxX {
    newFrame.origin.x = bounds.maxX - lastWindowFrame.width
//          |                |                  |
//        -868.0           -10.0              878.0
}
```

**Comment by T1mVo** [comment]
I will try a PR tomorrow. I will also have a look at the original problem of this issue.

**Comment by T1mVo** [comment]
I also found the cause of the original problem of this issue.

When the target frame of the preview is calculated, it stores its position to `LoopManager.lastTargetFrame`. When the targetFrame for the actual window is calculated, the position of the preview is used as the starting point.

**Comment by MrKai77** [comment]
> I also found the cause of the original problem of this issue.
> 
> When the target frame of the preview is calculated, it stores its position to `LoopManager.lastTargetFrame`. When the targetFrame for the actual window is calculated, the position of the preview is used as the starting point.

is this something you were planning on opening a new PR for?

Also just as a heads up, `LoopManager.lastTargetFrame` is only used to calculate size/position adjustments based on the *current* size/position, but is never used when using a window action such as `Right Half` etc. But if I understand correctly, you face this issue with all window actions right?

**Comment by T1mVo** [comment]
> is this something you were planning on opening a new PR for?

I will make a draft PR and see how far I can get.

> But if I understand correctly, you face this issue with all window actions right?

No, only when moving the window. But I guess the issue should also appear when resizing.
